### PR TITLE
Make `TaskExecutionListener` task agnostic

### LINKED ISSUE #18746: Generalize `TaskInputsListeners`
As the name suggests, `TaskInputsListeners` currently only inform the listeners about inputs for tasks. We should inform the listeners more generally about inputs for any kind of unit of work.

The current listeners are continuous build and the configuration cache.

---
cc: @gradle/execution

### PR DESCRIPTION
**Pull request by wolfs**
Instead of only capturing the inputs for tasks, we should capture inputs for all work items. This affects the kind of inputs continuous build and configuration caching detect.

Fixes #18746.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/internal/service/scopes/ExecutionGradleServices.java [hunk 1]
```diff
@@ -170,7 +171,7 @@ public ExecutionEngine createExecutionEngine(
             new LoadPreviousExecutionStateStep<>(
             new MarkSnapshottingInputsStartedStep<>(
             new RemoveUntrackedExecutionStateStep<>(
-            new SkipEmptyWorkStep(outputChangeListener, skipEmptyWorkOutputsCleanerSupplier,
+            new SkipEmptyWorkStep(outputChangeListener, workInputListeners, skipEmptyWorkOutputsCleanerSupplier,
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/internal/service/scopes/ExecutionGradleServices.java:174]: `SkipEmptyWorkStep` does too much. We don't necessarily need to fix this in this PR, but we should at some point separate the broadcasting of inputs from it.

File: subprojects/launcher/build.gradle.kts [hunk 2]
```diff
@@ -17,6 +17,7 @@ dependencies {
     implementation(project(":files"))
     implementation(project(":file-collections"))
     implementation(project(":snapshots"))
+    implementation(project(":execution"))
```
Comment (lptr) [inline subprojects/launcher/build.gradle.kts:20]: Why is this needed now?

File: subprojects/execution/build.gradle.kts [hunk 3]
```diff
@@ -11,6 +11,9 @@ dependencies {
     implementation(project(":core-api"))
     implementation(project(":functional"))
     implementation(project(":files"))
+    api(project(":file-collections")) {
+        because("WorkInputListener exposes FileCollectionInternal")
+    }
```
Comment (lptr) [inline subprojects/execution/build.gradle.kts:16]: Let's try to avoid this. (See other comments.)

**Review by lptr — CHANGES_REQUESTED** [review]
So now we are triggering continuous build on input changes to artifact transforms (good), and to Kotlin scripts but not to Groovy scripts (not so good).

How do we plan to address this? I see a number of options:

1) accept this, and explain it in documentation, arguing that more continuous build coverage is better, even if it adds a bit of confusion,
2) disable it for Kotlin scripts for now until we add Groovy script compilation via the execution engine,
3) implement Groovy script compilation via the execution engine, and release it together with this PR.

What should we do?

If we don't disable it for script compilation (1 or 3), we should also add some integration tests to cover that use case.

In either case we should add an integration test to cover the artifact transform input change use case.

File: subprojects/launcher/build.gradle.kts (hunk 2 above)
Comment (wolfs) [reply subprojects/launcher/build.gradle.kts:20]: The definition of `WorkInputListener(s)` is in execution, and we are using that for continuous build.

**Comment by wolfs** [comment]
I added tests for build script changes and artifact transform parameter changes. We don't trigger in any of the two cases :(. I suppose this is fine for now.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/steps/SkipEmptyWorkStep.java [hunk 4]
```diff
@@ -245,10 +252,31 @@ public WorkValidationContext getValidationContext() {
...
+                    @Override
+                    public void visitInputFileProperty(String propertyName, InputFingerprinter.InputPropertyType type, InputFingerprinter.FileValueSupplier value) {
+                        if (!onlyPrimaryInputs || type == InputFingerprinter.InputPropertyType.PRIMARY) {
+                            visitor.accept((FileCollectionInternal) value.getFiles());
+                        }
+                    }
+                });
+            }
+        });
+    }
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/steps/SkipEmptyWorkStep.java:261]: I think this is better. We could even go further, and hide `work` and the enum both, and just pass an opaque visitor here that the callee can use to visit whatever we deem important here. We don't have to do it in this PR though. Perhaps something to do when we restructure input visiting to use parameter objects.

**Review by lptr — APPROVED** [review]
LGTM.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/steps/SkipEmptyWorkStep.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/steps/SkipEmptyWorkStep.java:261]: Yeah, we could hide more from the work here. We'd still need to expose the display name, though. Configuration caching also wanted to store the individual properties instead of a single file collection at some point, so that is also something to keep in mind when designing the opaque visitor.

**Comment by wolfs** [comment]
@bot-gradle test and merge

**Comment by wolfs** [comment]
@bot-gradle test and merge
